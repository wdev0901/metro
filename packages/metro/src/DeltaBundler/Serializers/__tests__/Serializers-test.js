/**
 * Copyright (c) 2015-present, Facebook, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+javascript_foundation
 * @format
 */

'use strict';

jest.mock('../../../node-haste/lib/toLocalPath');
jest.mock('../../../Assets');

const toLocalPath = require('../../../node-haste/lib/toLocalPath');

const CURRENT_TIME = 1482363367000;

describe('Serializers', () => {
  const OriginalDate = global.Date;
  const getDelta = jest.fn();
  const getDependenciesFn = jest.fn();
  const postProcessModules = jest.fn();
  let deltaBundler;
  let Serializers;

  const deltaResponse = {
    id: '1234',
    pre: new Map([[1, {type: 'script', code: 'pre;', id: 1, path: '/pre.js'}]]),
    post: new Map([[2, {type: 'require', code: 'post;', id: 2, path: '/p'}]]),
    delta: new Map([
      [3, {type: 'module', code: 'module3;', id: 3, path: '/3.js'}],
      [4, {type: 'module', code: 'another;', id: 4, path: '/4.js'}],
    ]),
    inverseDependencies: [],
    reset: true,
  };

  function setCurrentTime(time: number) {
    global.Date = jest.fn(() => new OriginalDate(time));
  }

  beforeEach(() => {
    Serializers = require('../Serializers');

    getDelta.mockReturnValueOnce(Promise.resolve(deltaResponse));
    getDependenciesFn.mockReturnValue(Promise.resolve(() => new Set()));
    postProcessModules.mockImplementation(modules => modules);

    deltaBundler = {
      getDeltaTransformer: jest.fn().mockReturnValue(
        Promise.resolve({
          getDelta,
          getDependenciesFn,
        }),
      ),
      getPostProcessModulesFn() {
        return postProcessModules;
      },
    };

    toLocalPath.mockImplementation((roots, path) => path.replace(roots[0], ''));

    setCurrentTime(CURRENT_TIME);
  });

  it('should return the stringified delta bundle', async () => {
    expect(
      await Serializers.deltaBundle(deltaBundler, 'foo', {deltaBundleId: 10}),
    ).toMatchSnapshot();

    // Simulate a delta with some changes now
    getDelta.mockReturnValueOnce(
      Promise.resolve({
        id: '1234',
        delta: new Map([[3, {code: 'modified module;'}], [4, null]]),
        pre: new Map(),
        post: new Map(),
        inverseDependencies: [],
      }),
    );

    expect(
      await Serializers.deltaBundle(deltaBundler, 'foo', {deltaBundleId: 10}),
    ).toMatchSnapshot();
  });

  it('should build the full JS bundle', async () => {
    expect(
      await Serializers.fullBundle(deltaBundler, {
        sourceMapUrl: 'http://localhost:8081/myBundle.js',
      }),
    ).toMatchSnapshot();

    getDelta.mockReturnValueOnce(
      Promise.resolve({
        id: '1234',
        delta: new Map([[3, {code: 'modified module;'}], [4, null]]),
        pre: new Map([[5, {code: 'more pre;'}]]),
        post: new Map([[6, {code: 'bananas;'}], [7, {code: 'apples;'}]]),
        inverseDependencies: [],
      }),
    );
    setCurrentTime(CURRENT_TIME + 5000);

    expect(
      await Serializers.fullBundle(deltaBundler, {
        sourceMapUrl: 'http://localhost:8081/myBundle.js',
      }),
    ).toMatchSnapshot();
  });

  it('should pass the sourcemapURL param to the transformer', async () => {
    await Serializers.fullBundle(deltaBundler, {
      sourceMapUrl: 'http://localhost:8081/myBundle.js',
    });

    expect(deltaBundler.getDeltaTransformer.mock.calls[0][1]).toEqual({
      deltaBundleId: '1234',
      sourceMapUrl: 'http://localhost:8081/myBundle.js',
    });
  });

  // This test actually does not test the sourcemaps generation logic, which
  // is already tested in the source-map file.
  it('should build the full Source Maps', async () => {
    expect(await Serializers.fullSourceMap(deltaBundler, {})).toMatchSnapshot();

    getDelta.mockReturnValueOnce(
      Promise.resolve({
        id: '1234',
        delta: new Map([[3, {code: 'modified module;'}], [4, null]]),
        pre: new Map([[5, {code: 'more pre;'}]]),
        post: new Map([[6, {code: 'bananas;'}], [7, {code: 'apples;'}]]),
        inverseDependencies: [],
      }),
    );
    setCurrentTime(CURRENT_TIME + 5000);

    expect(await Serializers.fullSourceMap(deltaBundler, {})).toMatchSnapshot();
  });

  it('should return all the bundle modules', async () => {
    expect(await Serializers.getAllModules(deltaBundler, {})).toMatchSnapshot();

    getDelta.mockReturnValueOnce(
      Promise.resolve({
        delta: new Map([[3, {code: 'modified module;'}], [4, null]]),
        pre: new Map([[5, {code: 'more pre;'}]]),
        post: new Map([[6, {code: 'bananas;'}], [7, {code: 'apples;'}]]),
        inverseDependencies: [],
      }),
    );

    expect(await Serializers.getAllModules(deltaBundler, {})).toMatchSnapshot();
  });

  it('should return the RAM bundle info', async () => {
    expect(
      await Serializers.getRamBundleInfo(deltaBundler, {}),
    ).toMatchSnapshot();

    getDelta.mockReturnValueOnce(
      Promise.resolve({
        delta: new Map([
          [3, {type: 'module', code: 'modified;', id: 3, path: '/foo/3.js'}],
          [7, {type: 'asset', code: 'code', id: 7, path: '/foo/path.png'}],
          [8, {type: 'module', code: 'code', id: 8, path: '/foo/8.js'}],
          [9, {type: 'asset', code: 'code', id: 9, path: '/foo/path3.png'}],
        ]),
        pre: new Map([
          [5, {type: 'script', code: 'more pre;', id: 5, path: '/foo/5.js'}],
        ]),
        post: new Map([
          [6, {type: 'require', code: 'bananas;', id: 6, path: '/foo/6.js'}],
          [10, {type: 'comment', code: 'bananas;', id: 10, path: '/foo/10.js'}],
        ]),
        inverseDependencies: [],
      }),
    );

    expect(
      await Serializers.getRamBundleInfo(deltaBundler, {}),
    ).toMatchSnapshot();
  });

  it('should use the preloadedModules and ramGroup configs to build a RAM bundle', async () => {
    getDelta.mockReset();
    getDependenciesFn.mockReset();

    getDelta.mockReturnValue(
      Promise.resolve({
        delta: new Map([
          [3, {type: 'module', code: 'code', id: 3, path: '/foo/3.js'}],
          [4, {type: 'module', code: 'code', id: 4, path: '/foo/4.js'}],
          [5, {type: 'module', code: 'code', id: 5, path: '/foo/5.js'}],
          [6, {type: 'module', code: 'code', id: 6, path: '/foo/6.js'}],
        ]),
        pre: new Map([
          [7, {type: 'script', code: 'more pre;', id: 7, path: '/foo/7.js'}],
        ]),
        post: new Map([
          [8, {type: 'require', code: 'bananas;', id: 8, path: '/foo/8.js'}],
        ]),
        inverseDependencies: [],
        reset: 1,
      }),
    );

    getDependenciesFn.mockReturnValue(
      Promise.resolve(path => {
        expect(path).toBe('/foo/5.js');

        return new Set(['/foo/6.js']);
      }),
    );

    const getTransformOptions = async () => ({
      preloadedModules: {'/foo/3.js': true},
      ramGroups: ['/foo/5.js'],
    });

    expect(
      await Serializers.getRamBundleInfo(deltaBundler, {}, getTransformOptions),
    ).toMatchSnapshot();
  });

  it('should post-process the modules', async () => {
    postProcessModules.mockImplementation(modules => {
      return modules.sort(
        (a, b) => +(a.path < b.path) || +(a.path === b.path) - 1,
      );
    });

    expect(await Serializers.getAllModules(deltaBundler, {})).toMatchSnapshot();
  });
});
