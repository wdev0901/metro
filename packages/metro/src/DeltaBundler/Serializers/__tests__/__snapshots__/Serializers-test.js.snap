// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Serializers should return the RAM bundle info 1`] = `
Object {
  "getDependencies": [Function],
  "groups": Map {},
  "lazyModules": Array [
    Object {
      "code": "module3;",
      "id": 3,
      "map": Object {
        "file": "/3.js",
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": undefined,
      "source": undefined,
      "sourcePath": "/3.js",
      "type": "module",
    },
    Object {
      "code": "another;",
      "id": 4,
      "map": Object {
        "file": "/4.js",
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": undefined,
      "source": undefined,
      "sourcePath": "/4.js",
      "type": "module",
    },
  ],
  "startupModules": Array [
    Object {
      "code": "pre;",
      "id": 1,
      "map": Object {
        "file": "/pre.js",
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": undefined,
      "source": undefined,
      "sourcePath": "/pre.js",
      "type": "script",
    },
    Object {
      "code": "post;",
      "id": 2,
      "map": Object {
        "file": "/p",
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": undefined,
      "source": undefined,
      "sourcePath": "/p",
      "type": "require",
    },
  ],
}
`;

exports[`Serializers should return the RAM bundle info 2`] = `
Object {
  "getDependencies": [Function],
  "groups": Map {},
  "lazyModules": Array [
    Object {
      "code": "modified;",
      "id": 3,
      "map": Object {
        "file": "/foo/3.js",
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": undefined,
      "source": undefined,
      "sourcePath": "/foo/3.js",
      "type": "module",
    },
    Object {
      "code": "another;",
      "id": 4,
      "map": Object {
        "file": "/4.js",
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": undefined,
      "source": undefined,
      "sourcePath": "/4.js",
      "type": "module",
    },
    Object {
      "code": "code",
      "id": 7,
      "map": Object {
        "file": "/foo/path.png",
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": undefined,
      "source": undefined,
      "sourcePath": "/foo/path.png",
      "type": "asset",
    },
    Object {
      "code": "code",
      "id": 8,
      "map": Object {
        "file": "/foo/8.js",
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": undefined,
      "source": undefined,
      "sourcePath": "/foo/8.js",
      "type": "module",
    },
    Object {
      "code": "code",
      "id": 9,
      "map": Object {
        "file": "/foo/path3.png",
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": undefined,
      "source": undefined,
      "sourcePath": "/foo/path3.png",
      "type": "asset",
    },
  ],
  "startupModules": Array [
    Object {
      "code": "pre;",
      "id": 1,
      "map": Object {
        "file": "/pre.js",
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": undefined,
      "source": undefined,
      "sourcePath": "/pre.js",
      "type": "script",
    },
    Object {
      "code": "more pre;",
      "id": 5,
      "map": Object {
        "file": "/foo/5.js",
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": undefined,
      "source": undefined,
      "sourcePath": "/foo/5.js",
      "type": "script",
    },
    Object {
      "code": "post;",
      "id": 2,
      "map": Object {
        "file": "/p",
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": undefined,
      "source": undefined,
      "sourcePath": "/p",
      "type": "require",
    },
    Object {
      "code": "bananas;",
      "id": 6,
      "map": Object {
        "file": "/foo/6.js",
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": undefined,
      "source": undefined,
      "sourcePath": "/foo/6.js",
      "type": "require",
    },
  ],
}
`;

exports[`Serializers should return the stringified delta bundle 1`] = `
Object {
  "bundle": "{\\"id\\":\\"1234\\",\\"pre\\":[[1,\\"pre;\\"]],\\"post\\":[[2,\\"post;\\"]],\\"delta\\":[[3,\\"module3;\\"],[4,\\"another;\\"]],\\"reset\\":true}",
  "numModifiedFiles": 4,
}
`;

exports[`Serializers should return the stringified delta bundle 2`] = `
Object {
  "bundle": "{\\"id\\":\\"1234\\",\\"pre\\":[],\\"post\\":[],\\"delta\\":[[3,\\"modified module;\\"],[4,null]]}",
  "numModifiedFiles": 2,
}
`;

exports[`Serializers should use the preloadedModules and ramGroup configs to build a RAM bundle 1`] = `
Object {
  "getDependencies": [Function],
  "groups": Map {
    5 => Set {
      6,
    },
  },
  "lazyModules": Array [
    Object {
      "code": "code",
      "id": 4,
      "map": Object {
        "file": "/foo/4.js",
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": undefined,
      "source": undefined,
      "sourcePath": "/foo/4.js",
      "type": "module",
    },
    Object {
      "code": "code",
      "id": 5,
      "map": Object {
        "file": "/foo/5.js",
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": undefined,
      "source": undefined,
      "sourcePath": "/foo/5.js",
      "type": "module",
    },
    Object {
      "code": "code",
      "id": 6,
      "map": Object {
        "file": "/foo/6.js",
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": undefined,
      "source": undefined,
      "sourcePath": "/foo/6.js",
      "type": "module",
    },
  ],
  "startupModules": Array [
    Object {
      "code": "more pre;",
      "id": 7,
      "map": Object {
        "file": "/foo/7.js",
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": undefined,
      "source": undefined,
      "sourcePath": "/foo/7.js",
      "type": "script",
    },
    Object {
      "code": "code",
      "id": 3,
      "map": Object {
        "file": "/foo/3.js",
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": undefined,
      "source": undefined,
      "sourcePath": "/foo/3.js",
      "type": "module",
    },
    Object {
      "code": "bananas;",
      "id": 8,
      "map": Object {
        "file": "/foo/8.js",
        "mappings": "",
        "names": Array [],
        "sources": Array [],
        "sourcesContent": Array [],
        "version": 3,
      },
      "name": undefined,
      "source": undefined,
      "sourcePath": "/foo/8.js",
      "type": "require",
    },
  ],
}
`;
